crates/red_knot/src/semantic.rs

```diff
@@ -267,7 +252,36 @@ impl PreorderVisitor<'_> for SemanticIndexer {
 
         self.expressions
             .insert(NodeKey::from_node(expr.into()), expression_id);
-        ast::visitor::preorder::walk_expr(self, expr);
+
+        match expr {
+            ast::Expr::Name(ast::ExprName { id, ctx, .. }) => {
+                let flags = match ctx {
+                    ast::ExprContext::Load => SymbolFlags::IS_USED,
+                    ast::ExprContext::Store => SymbolFlags::IS_DEFINED,
```

Comment:
Should we also immediately mark any symbol defined using a walrus as both defined and used?

Reply:
The walrus expression should record the symbol as defined, not as used; a use is a load of a name, not a store to it.

But we already do record it as defined! Because we visit `node.target` on the walrus, which will be a `Name` node with ctx `Store`, so we will recursively hit this `Name` case, with `current_definition` set to the walrus expression.

Reply:
> The walrus expression should record the symbol as defined, not as used; a use is a load of a name, not a store to it.

Right, but in e.g. `x = (y := 1) + 2`, you could argue that `y` is being defined (assigned to `1`) and then _immediately_ used (added to `2`) in the same breath?

> But we already do record it as defined! Because we visit `node.target` on the walrus, which will be a `Name` node with ctx `Store`, so we will recursively hit this `Name` case, with `current_definition` set to the walrus expression.

Ah gotcha, thanks!

Reply:
> Right, but in e.g. x = (y := 1) + 2, you could argue that y is being defined (assigned to 1) and then immediately used (added to 2) in the same breath?

Aha! Yes, that's an interesting point. You could look at it that way, but this is not really the way Python looks at it, in terms of the AST (there is no `Name` node for `y` with context `Load` here) or in terms of the compiled bytecode (there is no `LOAD_FAST/LOAD_NAME` of `y` here). For basically exactly the same reasons, it is not the simplest way for us to look at it, either. There is no "use" of `y` here because there is no need for us to look up `y` and resolve reachable definition(s) for it: we know that definition would always resolve to this very walrus expression, and when doing type inference we always already have the type of that expression in hand! So it's much simpler to look at it as simply "the type of the walrus expression is the type of its right-hand side, and as a side effect it also defines a name." This way of looking at it is how CPython handles it and how we handle it in this PR, and it doesn't require treating the walrus expression as a "use" of its target name.